### PR TITLE
[PatternMatch] Use `m_SpecificCmp` matchers. NFC.

### PR DESCRIPTION
**Pull request by dtcxzyw**
Compile-time improvement: http://llvm-compile-time-tracker.com/compare.php?from=13996378d81c8fa9a364aeaafd7382abbc1db83a&to=861ffa4ec5f7bde5a194a7715593a1b5359eb581&stat=instructions:u
baseline: 803eaf29267c6aae9162d1a83a4a2ae508b440d3
```
Top 5 improvements:
  stockfish/movegen.ll 2541620819 2538599412 -0.12%
  minetest/profiler.cpp.ll 431724935 431246500 -0.11%
  abc/luckySwap.c.ll 581173720 580581935 -0.10%
  abc/kitTruth.c.ll 2521936288 2519445570 -0.10%
  abc/extraUtilTruth.c.ll 1216674614 1215495502 -0.10%
Top 5 regressions:
  openssl/libcrypto-shlib-sm4.ll 1155054721 1155943201 +0.08%
  openssl/libcrypto-lib-sm4.ll 1155054838 1155943063 +0.08%
  spike/vsm4r_vv.ll 1296430080 1297039258 +0.05%
  spike/vsm4r_vs.ll 1312496906 1313093460 +0.05%
  nuttx/lib_rand48.c.ll 126201233 126246692 +0.04%
Overall: -0.02112308%
```

### CODE REVIEW

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 1]
```diff
@@ -110,10 +110,11 @@ static Value *foldSelectWithBinaryOp(Value *Cond, Value *TrueVal,
   // -->
   // %TV
   Value *X, *Y;
-  if (!match(Cond, m_c_BinOp(m_c_ICmp(Pred1, m_Specific(TrueVal),
-                                      m_Specific(FalseVal)),
-                             m_ICmp(Pred2, m_Value(X), m_Value(Y)))) ||
-      Pred1 != Pred2 || Pred1 != ExpectedPred)
+  if (!match(
+          Cond,
+          m_c_BinOp(m_c_ICmp(Pred1, m_Specific(TrueVal), m_Specific(FalseVal)),
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:115]: Shouldn't this be?
```suggestion
          m_c_BinOp(m_c_SpecificICmp(ExpectedPred, m_Specific(TrueVal), m_Specific(FalseVal)),
```

Comment (nikic) [reply llvm/lib/Analysis/InstructionSimplify.cpp:115]: Also remove the now dead variable declarations?

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 2]
```diff
@@ -1697,9 +1697,8 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
   ICmpInst::Predicate Pred;
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1697]: Remove variable or move it to where it is needed now?

**Comment by nikic** [comment]
> ```
> Top 5 improvements:
>   faiss/IndexIVFFastScan.cpp.ll 7502943608 7223027166 -3.73%
>   gromacs/pull.cpp.ll 3581024503 3451962626 -3.60%
>   spike/regnames.ll 496396386 481168580 -3.07%
>   lightgbm/gradient_discretizer.cpp.ll 1364403988 1324787616 -2.90%
>   meshlab/cube_style_precomputation.cpp.ll 29532646099 28707512827 -2.79%
> Top 5 regressions:
>   openjdk/ad_x86_gen.ll 18251379251 20870783067 +14.35%
>   faiss/IndexIVFAdditiveQuantizerFastScan.cpp.ll 825075373 857500006 +3.93%
>   brotli/encoder_dict.c.ll 543355291 560137438 +3.09%
>   gromacs/constr.cpp.ll 2099218140 2163924992 +3.08%
>   lightgbm/dataset_loader.cpp.ll 7317833652 7536019648 +2.98%
> Overall: -0.02132359%
> ```

I don't see how this change can cause any significant compile-time changes (unless it's due to the bug in foldSelectWithBinaryOp). Can you please check what kind of results you get if you run the compile-time test against a no-op change to see what the noise level looks like?

**Comment by dtcxzyw** [comment]
Can you please check what kind of results you get if you run the compile-time test against a no-op change to see what the noise level looks like?

Tested with 803eaf29267c6aae9162d1a83a4a2ae508b440d3

Test1 vs Test2
```
Top 5 improvements:
  openjdk/ad_x86_gen.ll 19142207641 18261250223 -4.60%
  lightgbm/metadata.cpp.ll 5027690725 4825166371 -4.03%
  lightgbm/metric.cpp.ll 7180151122 6963045147 -3.02%
  lightgbm/gradient_discretizer.cpp.ll 1382530605 1340875607 -3.01%
  gromacs/atomdata.cpp.ll 2513395847 2437906279 -3.00%
Top 5 regressions:
  lightgbm/gbdt.cpp.ll 5331964084 5518023064 +3.49%
  faiss/IndexIVFFlat.cpp.ll 1736529706 1796315649 +3.44%
  faiss/lattice_Zn.cpp.ll 2131208372 2186387362 +2.59%
  lightgbm/linear_tree_learner.cpp.ll 9769789458 10022534163 +2.59%
  gromacs/constr.cpp.ll 2078220850 2125769072 +2.29%
Overall: -0.00069448%
```
Test1 vs Test3
```
Top 5 improvements:
  faiss/IndexIVFFastScan.cpp.ll 7434390364 7077625815 -4.80%
  lightgbm/metadata.cpp.ll 5027690725 4827333025 -3.99%
  lightgbm/sample_strategy.cpp.ll 1047493568 1008880792 -3.69%
  lightgbm/gradient_discretizer.cpp.ll 1382530605 1337101867 -3.29%
  faiss/IndexHNSW.cpp.ll 2671482829 2589963057 -3.05%
Top 5 regressions:
  openjdk/ad_x86_gen.ll 19142207641 23882411724 +24.76%
  faiss/IndexIVF.cpp.ll 2380453173 2463393700 +3.48%
  faiss/AdditiveQuantizer.cpp.ll 1411960257 1451771018 +2.82%
  lightgbm/gbdt.cpp.ll 5331964084 5475595898 +2.69%
  faiss/MetaIndexes.cpp.ll 1164233270 1194513258 +2.60%
Overall: 0.00045434%
```
Test2 vs Test3
```
Top 5 improvements:
  faiss/IndexIVFFastScan.cpp.ll 7413814621 7077625815 -4.53%
  lightgbm/linear_tree_learner.cpp.ll 10022534163 9677397451 -3.44%
  lightgbm/voting_parallel_tree_learner.cpp.ll 48690920539 47030359898 -3.41%
  faiss/IndexIVFFlat.cpp.ll 1796315649 1740205247 -3.12%
  lightgbm/data_parallel_tree_learner.cpp.ll 3577363719 3505678627 -2.00%
Top 5 regressions:
  openjdk/ad_x86_gen.ll 18261250223 23882411724 +30.78%
  faiss/IndexIVF.cpp.ll 2330588660 2463393700 +5.70%
  faiss/IndexIVFAdditiveQuantizerFastScan.cpp.ll 825778559 848334894 +2.73%
  gromacs/pullutil.cpp.ll 1468623225 1506443771 +2.58%
  gromacs/freeenergydispatch.cpp.ll 726020265 740090114 +1.94%
Overall: 0.00114883%
```

**Comment by nikic** [comment]
Okay, so it looks like the top 5 improvements/regressions are very noisy. I did wonder why I wasn't able to reproduce all regressions on previous PRs.

To mostly avoid this problem I collect per-benchmark stddev data and only consider differences starting at 3-4 sigma significant.

**Comment by dtcxzyw** [comment]
> Okay, so it looks like the top 5 improvements/regressions are very noisy. I did wonder why I wasn't able to reproduce all regressions on previous PRs.
> 
> To mostly avoid this problem I collect per-benchmark stddev data and only consider differences starting at 3-4 sigma significant.

Oops, I forgot to disable ASLR after the last power outage in my lab :(
With ASLR off:
```
Top 5 improvements:
Top 5 regressions:
Overall: 0.00000013%
```

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**